### PR TITLE
Queries branch

### PR DESCRIPTION
**Pull request by NematMimi-22**

### CODE REVIEW

File: Seeds/BookSeed.sql [hunk 1]
```diff
@@ -0,0 +1,1002 @@
+﻿use Tech_Library; 
+
+insert into Books ([Title], [Author], [ISBN], [PublishedDate], [Genre], [ShelfLocation], [CurrentStatus]) values ('Thief Who Came to Dinner, The', 'Amy Smith', '712351913-2', '8/14/2005', 'non-fiction', '0904-6102', 'Available');
```
Comment (amal-r365) [inline Seeds/BookSeed.sql]: please bulk insert

File: CreatedTables.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
+CREATE TABLE Books (
```
Comment (amal-r365) [inline CreatedTables.sql]: 1. always create if not exist.
2. let's have a file for each table

File: Functions/Add_Borrowers.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CREATE PROC sp_AddNewBorrower
```
Comment (amal-r365) [inline Functions/Add_Borrowers.sql]: always create or alter

File: 1.ListofBorrowedBooks.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+use Tech_Library;
+SELECT * FROM Books JOIN Loans ON Books.BookID = Loans.BookID WHERE Loans.BorrowerID = 1;
```
Comment (AmalDarKhalil) [inline 1.ListofBorrowedBooks.sql]: please notice the columns, we only need book columns

File: 2.ActiveBorrowerswithCTEs.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+use Tech_Library;
+WITH BorrowerCount AS (
+    SELECT BorrowerID, COUNT(*) AS BorrowedCount
+    FROM Loans
+    WHERE DateReturned IS NULL
+    GROUP BY BorrowerID
+)
+SELECT *
```
Comment (AmalDarKhalil) [inline 2.ActiveBorrowerswithCTEs.sql]: again, be careful with select and only select target columns
this comment is applied to all files, you can return the full name here and the book name in the previous comment

File: 2.ActiveBorrowerswithCTEs.sql [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+FROM Borrowers B
+JOIN BorrowerCount BC ON B.BorrowerID = BC.BorrowerID
+WHERE BC.BorrowedCount >= 2;
```
Comment (AmalDarKhalil) [inline 2.ActiveBorrowerswithCTEs.sql]: let's check if we can add `HAVING COUNT(*) = 0` in L7 instead

File: 4.PopularGenreAnalysis.sql [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ BB.Genre,
+ BB.BorrowMonth,
+ BB.BorrowYear,
+ BB.GenreCount
+FROM
+    BorrowedBooks BB
+WHERE
+    GenreRank = 1 
+ORDER BY
+    BorrowMonth;
```
Comment (AmalDarKhalil) [inline 4.PopularGenreAnalysis.sql]: hmm, do we need this? we already returned only one raw, don't we?

File: 5.Stored ProcedureAdd New Borrowers.sql [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+CREATE PROC sp_AddNewBorrower 
+	@firstname NVARCHAR(64),
+	@lastname NVARCHAR(64),
+	@email NVARCHAR(256),
+	@birthdate DATE,
+	@membershipDate Date
+AS
+BEGIN
+
+	IF EXISTS (SELECT '*' FROM Borrowers WHERE Email = @email)
```
Comment (AmalDarKhalil) [inline 5.Stored ProcedureAdd New Borrowers.sql]:
```suggestion
	IF EXISTS (SELECT 1 1 FROM Borrowers WHERE Email = @email)
```

File: 6.CalculateOverdueFees.sql [hunk 9]
```diff
@@ -0,0 +1,24 @@
+DROP FUNCTION IF EXISTS fn_CalculateOverdueFees;
+GO
+CREATE FUNCTION fn_CalculateOverdueFees(@LoanID INT)
+RETURNS int
+AS
+BEGIN
+    DECLARE @OverdueFee int;
+    DECLARE @DueDate DATE;
+    DECLARE @DateReturned DATE;
+	DECLARE @DaysOverdue INT;
```
Comment (AmalDarKhalil) [inline 6.CalculateOverdueFees.sql]:
```suggestion
    DECLARE @DaysOverdue INT;
```

File: 6.CalculateOverdueFees.sql [hunk 10]
```diff
@@ -0,0 +1,24 @@
+DROP FUNCTION IF EXISTS fn_CalculateOverdueFees;
+GO
+CREATE FUNCTION fn_CalculateOverdueFees(@LoanID INT)
+RETURNS int
+AS
+BEGIN
+    DECLARE @OverdueFee int;
```
Comment (AmalDarKhalil) [inline 6.CalculateOverdueFees.sql]:
```suggestion
    DECLARE @OverdueFee INT;
```

File: 6.CalculateOverdueFees.sql [hunk 11]
```diff
@@ -0,0 +1,24 @@
+DROP FUNCTION IF EXISTS fn_CalculateOverdueFees;
+GO
+CREATE FUNCTION fn_CalculateOverdueFees(@LoanID INT)
+RETURNS int
```
Comment (AmalDarKhalil) [inline 6.CalculateOverdueFees.sql]:
```suggestion
RETURNS INT
```

File: 6.CalculateOverdueFees.sql [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    DECLARE @OverdueFee int;
+    DECLARE @DueDate DATE;
+    DECLARE @DateReturned DATE;
+	DECLARE @DaysOverdue INT;
+    SELECT @DueDate = DueDate, @DateReturned = DateReturned
+    FROM Loans
+    WHERE LoanID = @LoanID;
+    IF @DateReturned IS NULL
+        SET @DateReturned = GETDATE();
+    SET @DaysOverdue = DATEDIFF(DAY, @DueDate, @DateReturned);
```
Comment (AmalDarKhalil) [inline 6.CalculateOverdueFees.sql]:
```suggestion
        SET @DaysOverdue = DATEDIFF(DAY, @DueDate, @DateReturned);
```

File: 6.CalculateOverdueFees.sql [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    IF @DaysOverdue <= 30
+        SET @OverdueFee = @DaysOverdue; 
+    ELSE
+        SET @OverdueFee = 30 + @DaysOverdue * 2.00;
```
Comment (AmalDarKhalil) [inline 6.CalculateOverdueFees.sql]: We count the first 30 days as 1$ cost, but we want to eliminate them when calculating for over 30 days costs
Meaning: Fee = fee for 30 days +  (days overdue - 30) * cost for over 30 days
if you're late 31 days the fees will be 32$ and not 30 + 62 = 92$ 
```suggestion
        SET @OverdueFee = 30 + (@DaysOverdue - 30) * 2.00; 
```

File: 7.BookBorrowingFrequency.sql [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+CREATE FUNCTION fn_BookBorrowingFrequency
+(
+    @BookID INT
+)
+RETURNS INT
+AS
+BEGIN
+    DECLARE @BorrowingCount INT;
+    SELECT @BorrowingCount = COUNT(*) FROM Loans WHERE BookID = @BookID;
+    RETURN @BorrowingCount;
```
Comment (AmalDarKhalil) [inline 7.BookBorrowingFrequency.sql]: can't we call the count immediately without any extra storage?

File: 9.Author_Popularity.sql [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+    b.Author AS AuthorName,
+    COUNT(*) AS BorrowingFrequency
+FROM
+    Loans l
+ JOIN
+    Books b ON l.BookID = b.BookID
+GROUP BY
+    b.Author
+ORDER BY
+    BorrowingFrequency ASC;
```
Comment (AmalDarKhalil) [inline 9.Author_Popularity.sql]: let's order by the most popular ones first

**Pull request merged**